### DISCUSSION TITLE
Is it possible to make requests "as" a particular user on the server?

### DISCUSSION THREAD

**Discussion by churichard** [supabase/supabase#1094] [Questions] [answered]
For context, I'm using Next.js with supabase-js, roughly following the [Next.js with supabase auth example](https://github.com/supabase/supabase/tree/master/examples/nextjs-with-supabase-auth).

I have a policy `uid() = user_id` on a table with row level security turned on, and I want it to work nicely with requests happening on the server side as well. Right now what happens is that requests on the server side fail because I'm using supabase-js with the anon key, and the server doesn't "know" what user is logged in.

I know it'll work if I use the `service_role` key, but I don't want to do that. My logic is in `getServerSideProps` and it would be very easy for someone to accidentally use the `service_role` key to make requests on the client side.

So, my question is: is it possible to "force" supabase-js to make requests *as* a certain user on the server? I can already get the user from the cookie using `supabase.auth.api.getUserByCookie` - can I somehow pass this into the supabase-js client so that it makes requests as that user?

**Comment by atnegrete** [comment]
Check out https://github.com/supabase/supabase/discussions/1067 

Can the access token be used on the server? I'm curious, following this too

**Reply by churichard** [reply]
Yeah, I saw that thread, but my question is a bit different.

In my case, I already know who the user is, and I just simply want to make the request as if I'm the user.

I noticed this line in the Authentication docs:

```
If no Authorization header is included, the API will assume that you are making a request with an anonymous user.

If an Authorization header is included, the API will "switch" to the role of the user making the request. See the User Management section for more details.
```

So it seems like I just need to include the Authorization header with the user access token. I can probably do this with a REST API call, but it would be nice if I could also do it using supabase-js as well.

**Reply by atnegrete** [reply]
@churichard  It looks like the issue may be here: 
https://github.com/supabase/postgrest-js/blob/7be73b7d2584ebe2725014a42050169f5bb0a62a/src/PostgrestClient.ts#L41

I suppose you could create a new postgrest client, and set the auth token there manually? I'm still getting familiarized with the source code though

**Comment by inian** [comment, marked as answer]
Hi @churichard, yes you can do that. You need to generate a JWT with the `sub` claim as the user id and sign the JWT with your JWT secret of your project. Then you can pass that jwt into supabase-js like this. 

```js
const supabase = createClient(supabaseUrl, supabaseKey, {
  headers: {
    apiKey: anonKey,
    Authorisation: `Bearer ${jwt}`
  }
});
```

**Reply by laznic** [reply]
Does this work properly? I've looked into the source code and `supabase-js` ([_getAuthHeaders](https://github.com/supabase/supabase-js/blob/e632f79156cc6e499db62e0f6b3e861856c90489/src/SupabaseClient.ts#L170)) always uses the API key as the Authorization header value, and therefore always sending anonymous requests. I guess this works when using the `supabase-js` client in the browser, however not in server side.

**Reply by easy-street** [reply]
@laznic, doesn't work for me, and i reach the same conclusion looking at the source: i don't see any way it could work.

i'm aware of two approaches with similarly funky ergonomics:

`supabase.auth.session`:

```
import { supabase } from "../utils/initSupabase";

export default function Foos({ foos }) {
  return <>{JSON.stringify(foos)}</>;
}

export async function getServerSideProps({ req }) {
  supabase.auth.session = () => ({ access_token: req.cookies["sb:token"] });

  return {
    props: {
      foos: await supabase.from("foos").select(),
    },
  };
}
```

and `PostgrestClient`:

```
import { PostgrestClient } from "@supabase/postgrest-js";

const postgrest = new PostgrestClient(
  `${process.env.NEXT_PUBLIC_SUPABASE_URL}/rest/v1`,
  {
    headers: { apikey: process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY },
  }
);

export default function Foos({ foos }) {
  return <>{JSON.stringify(foos)}</>;
}

export async function getServerSideProps({ req }) {
  return {
    props: {
      foos: await postgrest.auth(req.cookies["sb:token"]).from("foos").select(),
    },
  };
}
```

**Reply by brattonross** [reply]
I've tried something based off of the suggestion by @inian and also not had any luck in getting it to work. 

When I receive a request on the server I am taking the `sb:token` cookie sent by the client and using that to set the `Authorization` header on the server-side supabase client. I've assumed that this is the correct token to use as I can decode the JWT stored in that cookie and it contains the values I would expect.

**Reply by inian** [reply]
Not sure if it helps, but [this](https://github.com/supabase/storage-api/blob/3bc0aec42c2bc88cf533f1646bc9e8f627033455/src/routes/bucket/getBucket.ts#L38) is how the storage-api makes request to the database as a user.

**Reply by brattonross** [reply]
Thanks, that does in understanding how the API is interpreting the request.

However as others have pointed out, I think there is still the issue that the `SupabaseClient` itself will only ever send anonymous requests if used outside of a browser.

In my case I am doing the equivalent of the following in a Next.js API route:

```js
import { createClient } from '@supabase/supabase-js'

const supabase = createClient(url, anonKey, {
  headers: {
    apiKey: anonKey,
    Authorization: `Bearer ${jwt}`
  }
})

const result = await supabase.storage.from('avatars').download(avatarURL)
```

If I trace back through the call stack, I can see that the `download()` call is getting its headers from the headers passed to the `SupabaseStorageClient` constructor, and that comes from [this line](https://github.com/supabase/supabase-js/blob/master/src/SupabaseClient.ts#L74) in the `SupabaseClient` class.

The `_getAuthHeaders` method is supplying the headers that will ultimately be used in the API call to the Storage API (if I'm understanding it correctly), and the `Authorization` header is always calculated by first checking the current session to see if there is an existing `access_token`, or otherwise just using the anonymous key. Since we're on the server-side there won't be a session, so we always end up using the anon key, and therefore we'll always make an anonymous call, regardless of the headers passed in when creating the client.

**Reply by inian** [reply]
Yup, you are right. Can you create an issue in supabase-js for this? As a workaround, you can use postgrest-js (which supabase-js uses internally).

**Reply by easy-street** [reply]
keep an eye on the work in progress here: https://github.com/supabase/gotrue-js/pull/80

the proposed implementation there would unlock something like:
```
const supabase = createClient(url, anonKey);
supabase.auth.setSession(jwt);

const result = await supabase...
```
(may end up as `supabase.auth.setUser` instead, TBD)

**Reply by brattonross** [reply]
Thanks @inian, I've created a new issue which you can find [here](https://github.com/supabase/supabase-js/issues/166)

**Reply by inian** [reply]
That is a very good issue report @brattonross. Thanks ❤️

**Reply by easy-street** [reply]
> (may end up as `supabase.auth.setUser` instead, TBD)

following up: it actually ended up as `supabase.auth.setAuth`, now available in [v1.11.10](https://github.com/supabase/supabase-js/releases/tag/v1.11.10)

```
const supabase = createClient(url, anonKey);
supabase.auth.setAuth(jwt);

const result = await supabase...
```

**Reply by bennettdams** [reply]
@easy-street  This means that we have to create a Supabase client on every request, right? It would be nice if using the JWT of that "user-action" would be a function call or another parameter when executing something, instead of only when creating the client.

**Reply by lintonye** [reply]
Creating a Supabase client on every request doesn't sound very efficient. I wonder how much overhead creating a client would have. Is there a limit of the max number of clients?

**Reply by elanzini** [reply]
Might be a silly question but I could not find the answer here or https://github.com/supabase/supabase/discussions/6723. How do you get the jwt on the client side?

Is it the `access_token` inside local storage?

**Reply by csalmeida** [reply]
Hi all, 

I am trying to run a script that inserts a bunch of records into multiple tables programmatically but requires a user to be authenticated in some cases for all triggers to run correctly.

I think the `setAuth` method is no longer available to authenticate the request in a Node script.
Any thoughts on how to authenticate a user in a script would be very helpful! 🙏

**Reply by bombillazo** [reply]
Hey @csalmeida so far the solution we have is to create a supabase client instance and pass the `Authorization` header to the client using `option.global.headers`

**Reply by Jonatthu** [reply]
@bombillazo I did exactly that but still const session = await ctx.supabase.auth.getSession(); return null

**Reply by pierroo** [reply]
I'm a bit late but since this issue is still valid:
What about this "Creating a Supabase client on every request doesn't sound very efficient"?
Would there be any issue having to re-create the client each time?
did you get an answer to that @lintonye ?

**Reply by AlexIsMaking** [reply]
> Hey @csalmeida so far the solution we have is to create a supabase client instance and pass the Authorization header to the client using option.global.headers

In case anyone else is looking, here's the [docs](https://supabase.com/docs/reference/javascript/v1/upgrade-guide?queryGroups=version&version=2.0x#use-a-custom-accesstoken-jwt-with-supabase) for this.

**Comment by Abhimanyu08** [comment]
If using typescript you'll also need to set `token_type` and `user` on the object returned by `supabase.auth.session`:

```
export const getServerSideProps: GetServerSideProps = async (context) => {

	const { user } = await supabase.auth.api.getUserByCookie(context.req);
	supabase.auth.session = () => ({
		access_token: context.req.cookies["sb-access-token"] as string,
		token_type: "bearer",
		user,
	});

       return {
           props: {posts: supabase.from('posts').select()}
   } 

```

**Comment by bennettdams** [comment]
## Follow up for Supabase JS `v2`

You can now easily do this via the new `createServerSupabaseClient` API. It can either be used in `getServerSideProps` or in an API endpoint. With that client, you can do things on behalf of the user (like using the storage or getting his authentication).

[Here's an example](https://supabase.com/docs/guides/auth/auth-helpers/nextjs#server-side-rendering-ssr):

```ts
// Create authenticated Supabase Client
const supabase = createServerSupabaseClient(ctx);

// Check if we have a session
const {
  data: { session },
} = await supabase.auth.getSession();
```

I've built myself a helper that I now use everywhere:

```ts
export async function getUserFromRequest(
  req: NextApiRequest,
  res: NextApiResponse
): Promise<
  | {
      userId: string;
      error: null;
    }
  | {
      userId: null;
      error: Error;
    }
> {
  const supabaseServerClient = createSupabaseClientFromRequest(req, res);

  const userResult = await supabaseServerClient.auth.getUser();

  if (userResult.error || !userResult) {
    const msg = "Error while getting user.";
    console.error(msg, userResult.error);

    return {
      userId: null,
      error: new Error(`${msg} ${userResult.error.message}`),
    };
  } else {
    return { userId: userResult.data.user.id, error: null };
  }
}

```

**Reply by lgtoledo** [reply]
Can you give us github repo with the details of a working example please? Thank you in advance! (the function you've built)

**Reply by farzd** [reply]
what is ```createSupabaseClientFromRequest``` ?

**Reply by juanbercoff** [reply]
Another options is to modify the headers in the middleware and set the user-id, or access token as a new header and then consume it on your api endpoint.
```
export async function middleware(req: NextRequest) {
	const res = NextResponse.next();

	const supabase = createMiddlewareSupabaseClient({ req, res });

	const {
		data: { session },
	} = await supabase.auth.getSession();

	if (session) {
		const requestHeaders = new Headers(req.headers);
		requestHeaders.set('user-id', session.user.id);
		const response = NextResponse.next({
			request: {
				// New request headers
				headers: requestHeaders,
			},
		});
		return response;
	}
```

Either way, the approach of @bennettdams is preferred. It's in the documentation, and you don't have to type the headers.
https://supabase.com/docs/guides/auth/auth-helpers/nextjs

**Reply by juanbercoff** [reply]
You could also create a branded type to narrow the return value from getUserFromRequest.
```
type ValidUser = {
	userId: string;
	error: null;
};

type InvalidUser = {
	userId: null;
	error: Error;
};

export function isValidUser(
	userResponse: ValidUser | InvalidUser
): userResponse is ValidUser {
	return !userResponse.error; //or return true;
}
```

**Reply by geoffreygarrett** [reply]
This is *definitely* not something written with posterity on the horizon.
